### DISCUSSION TITLE
PID controller emergency off

### DISCUSSION THREAD

**Discussion by blacknell** [arendst/Tasmota#19314] [Q&A]
The PID controller used for heating control is far far superior than the thermostat controller in my experience. Without much work, I've seen that it is easily tuned and hunts the set point really well. The thermostat logic is quite complex and it's been difficult to fathom how to tune it for my heating.

One thing PID lacks is something akin to the thermostat's `StateEmergencySet` command which shuts it down and doesn't allow any on commands untlil either reboot or unsetting it.

I'd like to add this to PID but I wanted to check first. The PID controller is more generic aimed at attaining any measurement not just temperature. That said I do think it would promote use if we added.

I propose

- adding a command to switch it into emergency mode
- the behaviour and defaults would be such that not using it has no effect (of course)
- a reboot would unset the mode
- switching on power either auto or manual would have no effect when emergency mode is on

We could call it `ShutdownMode` rather than `EmergencyMode` if people prefer not associating it with an actual 'emergency'.

I don't believe this capability should be handled in home assistant because it needs to operate autonomously.

FYI. I use a DS18B20 to measure the internal temperature of the radiator and use a rule to trigger the emergency.

**Comment by Wintespe** [comment]
Combine PidAuto and  PidManualPower within the PID controller to shut down.


<html>
<head>

Command | Def | Description
-- | -- | --
PidPv | - | This Is the current actual 			process value.
PidSp | 19.5 | Setpoint value. This is the 			process value that the process is aiming for.
PidPb | 5 | Proportional band in process 			units (eg degrees). This controls the gain of the loop and is the 			range of process value over which the power output will go from 0 			to full power. The units are that of the process and setpoint, so 			for example in a heating application it might be set to 1.5 			degrees. 			Set 			to zero to implement a 2-step controller.
PidTi | 1800 | This is a setting for the 			integral time, in seconds. It represents the time constant of the 			integration effect. The larger the value the slower the integral 			effect will be. Obviously the slower the process is the larger 			this should be. For example for a domestic room heated by 			convection radiators a setting of one hour might be appropriate 			(in seconds). To disable the integral effect set this to a large 			number,
PidTd | 15 | This is a setting for the 			derivative time, in seconds. It represents the time constant of 			the derivative effect.The larger the value the greater will be the 			derivative effect. Typically this will be set to somewhat less 			than 25% of the integral setting, once the integral has been 			adjusted to the optimum value. To disable the derivative effect 			set this to 0. When initially tuning a loop it is often sensible 			to start with derivative zero and wind it in once other parameters 			have been setup.
PIDInitint | 0.5 | This is an initial value 			which is used to preset the integrated error value when the flow 			is deployed in order to assist in homing in on the setpoint the 			first time. It should be set to an estimate of what the power 			requirement might be in order to maintain the process at the 			setpoint. For example for a domestic room heating application it 			might be set to 0.2 indicating that 20% of the available power 			might be required to maintain the setpoint. The value is of no 			consequence apart from device restart.
PidMaxInterval | 300 | This is 			the maximum time in seconds that is expected between samples. It 			is provided to cope with unusual situations such as a faulty 			sensor that might prevent the node from being supplied with a 			process value. If no new process value is received for 			PidMaxInterval 			then the power is set to the value defined by 			 PidManualPower. 			Set to zerro 			if not used. In this case the controller 			is always 			activated.
PidDSmooth | 3 | IIn 			situations where the process sensor has limited resolution (e.g. 			DS18B20), using the derivative can be problematic when the process 			is changing slowly and the steps in the value cause spikes in the 			derivative. To reduce the effect PidDSmoothr 			can be set to apply a filter to the derivative term. It has been 			found that with the DS18B20 a value of 3 can be beneficial here, 			effectively providing a low pass filter for the derivative at 1/3 			the derivative time. This feature may also be useful if the 			process value is particularly noisy. The smaller the value the 			greater the filtering effect but the more it will reduce the 			effectiveness of the derivative. A value of zero disables this 			feature.
PidAuto | 1 | Auto mode 1 or 0 (for 			manual).This can be used to enable or disable the control 			(1=enable, auto mode, 0=disabled, manual mode). When in manual 			mode the output is set the value defined for PidManualPower
PidManualPower | 0 | Power output when in manual 			mode or fallback mode if too long elapses between process values.
PidUpdateSecs | 0 | Specifies 			the 			interval time in seconds  to 			execute the pid algorithm. 			If 			PidUpdateSecs = 0 the algorithm isonly 			triggerd each time a new temperature value is received. This time 			is specified by TELE_PERIOD. 			A 			value 			of 15 seconds may well be 			appropriate for a domestic room heating application.

</body>
</html>

**Reply by blacknell** [reply]
Yes, but this new method prevents switching off such as in a failure mode and is similar to that already in thermostat driver
